mountpoint-s3/src/inode.rs

```diff
@@ -360,6 +360,68 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Unlink the entry described by `parent_ino` and `name`.
+    ///
+    /// If the entry exists, delete it from S3 and the superblock.
+    pub async fn unlink<OC: ObjectClient>(
```

Comment:
In my implementation of `rmdir`, I am adding a `Deleted` state in `WriteStatus` in order to mark the inode as deleted so that any other concurrent operation on the given inode shall fail. Maybe it can work here as well? (Still not sure though if thats right approach for rmdir as well)

Reply:
I had a similar idea - I'm not sure if its the right approach right now, but we can always change it.

There's two different issues I suppose:
- Trying to perform a concurrent operations _during_ the `unlink`/`rmdir`
- Trying to perform a concurrent operation which is occuring after the `unlink`/`rmdir` is complete - i.e. the operation started and has the `Inode`, but now the other concurrent operations already completed.

Reply:
After reviewing how the VFS is handling concurrent `unlink`, I've moved the locking to take write/exclusive locks on both the parent and the child in line with the VFS implementation.

At the moment, we gain nothing but complexity by trying to reduce the time we hold locks inside Mountpoint.

Discussed a bit more in the `rmdir` PR too: https://github.com/awslabs/mountpoint-s3/pull/247#discussion_r1199074008